doc/man/gnunet.conf.5
.\" -*- mode: nroff -*-
.TH GNUNET.CONF "5" "October 26, 2018" "GNUnet"
.SH NAME
gnunet.conf \- GNUnet configuration file
.SH SYNOPSIS
~/.config/gnunet.conf
.SH DESCRIPTION
A GNUnet setup typically consists of a set of service processes run by a user
"gnunet" and a set of user-interface processes run by a standard account.
The default location for the configuration file for the services is
"~gnunet/.config/gnunet.conf"; however, as normal users also may need
read-access to this configuration, you might want to instead put the service
process configuration in "/etc/gnunet.conf".
gnunet\-setup (part of the GNUnet GTK package) can be used to edit this
configuration.  The parts of GNUnet that are run as a normal user may have
config options too and they read from "$HOME/.config/gnunet.conf".
The latter config file can skip any options for the services.
.PP
The basic structure of the configuration file is the following.  The file is
split into sections.  Every section begins with "[SECTIONNAME]" and contains
a number of options of the form "OPTION=VALUE".
Empty lines and lines beginning with a "#" are treated as comments.
Almost all options are optional and the tools resort to reasonable defaults
if they are not present.
.PP
Default values for all of the options can be found in the files in the
"$GNUNET_PREFIX/share/gnunet/config.d/" directory. A typical setup will
work out of the box with those. See the examples section below for
some common setups on top of that.
.SH General OPTIONS
Many options will be common between sections. They can be repeated under
each section with different values.  The "[PATHS]" section is special.
Here, it is possible to specify values for variables like "GNUNET_HOME".
Then, in all filenames that begin with "$GNUNET_HOME" the "$GNUNET_HOME"
will be replaced with the respective value at runtime.  The main use of
this is to redefine "$GNUNET_HOME", which by default points to "$HOME/.config/".
By setting this variable, you can change the location where GNUnet stores
its internal data.
gnunet.conf accepts the variable "GNUNET_TMP" which we suggest to use in
place of the absolute definition of "/tmp".
So instead of "/tmp/foo" you would write "$GNUNET_TMP/foo".
 The usage of "$GNUNET_TMP/foo", will result in "$TMPDIR/gnunet/foo", or
 "$TMP/gnunet/foo" and finally, if "TMPDIR" is undefined, "/tmp/gnunet/foo".
.PP
The following options are generic and shared by all services:
.IP HOSTNAME
    The hostname specifies the machine on which the service is running.
    This is usually "localhost".
.IP BINARY
    The filename that implements the service. For example "gnunet-service-ats".
.IP IMMEDIATE_START
    Start the service always when the peer starts.  Set to YES for services
    that should always be launched, even if no other service explicitly needs
    them.
.IP START_ON_DEMAND
    Set to YES to automatically start the service when it is requested by
    another service. YES for most GNUnet services.
.IP NOARMBIND
    Set to YES to never have ARM bind to the respective socket. This option is
    mostly for debugging in situations where ARM cannot pass the pre-bound
    socket to the child due to interference from PREFIX-commands.
    This option is only effective in combination with IMMEDIATE_START being YES.
    NO by default.
.IP PREFIX
    PREFIX the given command (with its arguments) to the actual BINARY to be
    executed. Useful to run certain services under special supervisors (like
    strace or valgrind).  Typically used in combination with IMMEDIATE_START
    and NOARMBIND. Empty by default.
.IP ACCEPT_FROM
    A semi-column separated list of IPv4 addresses that are allowed to use
    the service; usually 127.0.0.1.
.IP ACCEPT_FROM6
    A semi-column separated list of IPv6 addresses that are allowed to use the
    service; usually ::1.
.IP UNIXPATH
    Path to use for the UNIX domain socket for inter process communication with
    the service on POSIX systems.
.IP UNIX_MATCH_UID
    If UNIX domain sockets are used, set this to YES if only users with the same
    UID are allowed to access the service.
.IP UNIX_MATCH_GID
    If UNIX domain sockets are used, set this to YES if only users with the same
    GID are allowed to access the service.
.IP RUN_PER_USER
    Set to YES if this service should be run per-user, NO if this is a system
    service.  End-users should never have to change the defaults GNUnet provides
    for this option.
.SH ATS Options
.IP UNSPECIFIED_QUOTA_IN
    quotes in KiB or MiB per seconds.  Or use the word "unlimited"
.IP UNSPECIFIED_QUOTA_OUT
    quotes in KiB or MiB per seconds.  Or use the word "unlimited"
.IP LOOPBACK_QUOTA_IN
    quotes in KiB or MiB per seconds.  Or use the word "unlimited"
.IP LOOPBACK_QUOTA_OUT
    quotes in KiB or MiB per seconds.  Or use the word "unlimited"
.IP LAN_QUOTA_IN
    quotes in KiB or MiB per seconds.  Or use the word "unlimited"
.IP LAN_QUOTA_OUT
    quotes in KiB or MiB per seconds.  Or use the word "unlimited"
.IP WAN_QUOTA_IN
    quotes in KiB or MiB per seconds.  Or use the word "unlimited"
.IP WAN_QUOTA_OUT
    quotes in KiB or MiB per seconds.  Or use the word "unlimited"
.IP WLAN_QUOTA_IN
    quotes in KiB or MiB per seconds.  Or use the word "unlimited"
.IP WLAN_QUOTA_OUT
    quotes in KiB or MiB per seconds.  Or use the word "unlimited"
.SH EXAMPLES
This example is a simple way to get started, using a server that has a known
list of peers to get you started. Most users will be behind a firewall on
IPv4, as such NAT is enabled.  Please rememeber to change your IP address
to the actual external address for your usage.
.PP
    [hostlist]
    OPTIONS = \-b
    SERVERS = http://v9.gnunet.org:58080/

    [nat]
    BEHIND_NAT = YES
    ENABLE_UPNP = YES
    DISABLEV6 = YES
    EXTERNAL_ADDRESS = 157.166.249.10

    [arm]
    START_SYSTEM_SERVICES = YES
    START_USER_SERVICES = NO
.SH FILES
.TP
~/.config/gnunet.conf
GNUnet configuration file
.SH BUGS
Report bugs by using Mantis <https://bugs.gnunet.org/> or by sending
electronic mail to <bug-gnunet@gnu.org>
.SH SEE ALSO
\fBgnunet\-setup\fP(1), \fBgnunet\-arm\fP(1)
.PP
The full documentation for
.B gnunet
is maintained as a Texinfo manual.
If the
.B info
and
.B gnunet
programs are properly installed at your site, the command
.IP
.B info gnunet
.PP
should give you access to the complete handbook,
.IP
.B info gnunet-c-tutorial
.PP
will give you access to a tutorial for developers.
.PP
Depending on your installation, this information is also
available in
\fBgnunet\fP(7) and \fBgnunet-c-tutorial\fP(7).


ChangeLog
Thu Feb 28 19:07:59 CET 2019
  CADET: Fix in order, unrealiable message ordering

Thu 28 Feb 2019 10:54:58 AM CET
	Releasing GNUnet 0.11.0. -CG

Wed Jun  6 08:28:27 CEST 2018
	Releasing GNUnet 0.11.0pre66. Let's try to keep a better
	ChangeLog again henceforth. -CG

Sat Jan  7 16:41:34 CET 2017
	Converting NAT library to new NAT service (and
	splitting of nat-auto service for auto-configuration). -CG


contrib/services/shepherd/ng0_wip/gnunet.scm
;;; plant ---
;;; Copyright (C) 2016, 2017, 2018 Nils Gillmann <gillmann@infotropique.org>
;;;
;;; This file is part of plant.
;;;
;;; plant is free software; you can redistribute it and/or modify it
;;; under the terms of the GNU General Public License as published by
;;; the Free Software Foundation; either version 3 of the License, or (at
;;; your option) any later version.
;;;
;;; plant is distributed in the hope that it will be useful, but
;;; WITHOUT ANY WARRANTY; without even the implied warranty of
;;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;;; GNU General Public License for more details.
;;;
;;; You should have received a copy of thye GNU General Public License
;;; along with plant.  If not, see <http://www.gnu.org/licenses/>.

(define-module (infotropique services networking)
  #:use-module (infotropique services)
  #:use-module (infotropique services shepherd)
  #:use-module (infotropique services dbus)
  #:use-module (infotropique system shadow)
  #:use-module (infotropique system pam)
  #:use-module (infotropique packages admin)
  #:use-module (infotropique packages connman)
  #:use-module (infotropique packages linux)
  #:use-module (infotropique packages tor)
  #:use-module (infotropique packages messaging)
  #:use-module (infotropique packages networking)
  #:use-module (infotropique packages ntp)
  #:use-module (infotropique packages wicd)
  #:use-module (infotropique packages gnome)
  #:use-module (infotropique packages gnunet)
  #:use-module (plant gexp)
  #:use-module (plant records)
  #:use-module (plant modules)
  #:use-module (srfi srfi-1)
  #:use-module (srfi srfi-9)
  #:use-module (srfi srfi-26)
  #:use-module (ice-9 match)
  #:export (gnunet-configuration
            gnunet-configuration?
            gnunet-service
            gnunet-service-type))

;;;
;;; Commentary:
;;; gnunet (GNUnet) related services, mainly gnunet itself.
;;;

;; GENTOO OpenRC:
DONE: depends on "net".
DONE: PIDFILE=/run/gnunet/arm-service.pid
SUID_ROOT_HELPERS=exit, nat-server, nat-client, transport-bluetooth, transport-wlan, vpn

/var/lib/gnunet/.local/share/gnunet/gnunet.conf must be chmod 600 and chown gnunet:gnunet
/var/lib/gnunet/.cache/gnunet must exist.
/usr/lib/gnunet/libexec/gnunet-helper-SUID_ROOT_HELPERS must be s+u (--> suid)

/usr/lib/gnunet/libexec/gnunet-helper-dns must be: chown root:gnunetdns and chmod 4750
/usr/lib/gnunet/libexec/gnunet-service-dns must be: chown gnunet:gnunetdns and chmod 2750

directory with PID file must then be chowned by gnunet:gnunet

user gnunet startet dann /usr/lib/gnunet/libexec/gnunet-service-arm -d

stop process hat:
start-stop-daemon --stop --signal QUIT --pidfile ${PIDFILE}
sleep 1
killall -u gnunet
sleep 1
rm -rf /tmp/gnunet-gnunet-runtime >/dev/null 2>&1
rm -rf /tmp/gnunet-system-runtime >/dev/null 2>&1

/etc/nsswitch.conf kriegt den eintrag:
hosts:       files gns [NOTFOUND=return] dns

und die dateien die in der source rumliegen bzgl nss müssen noch kopiert werden
UND nss muss sie finden.



(define-record-type* <gnunet-configuration>
  gnunet-configuration make-gnunet-configuration
  gnunet-configuration?
  (package           gnunet-configuration-package
                     (default gnunet))
  (config-file       gnunet-configuration-config-file
                     (default %default-gnunet-config-file)))

;; TODO: [PATHS] DEFAULTCONFIG = ?
(define %default-gnunet-config-file
  (plain-file "gnunet.conf" "
[PATHS]
SERVICEHOME = /var/lib/gnunet
GNUNET_CONFIG_HOME = /var/lib/gnunet

[arm]
SYSTEM_ONLY = YES
USER_ONLY = NO

[nat]
BEHIND_NAT = YES
ENABLE_UPNP = NO
USE_LOCALADDR = NO
DISABLEV6 = YES

[hostlist]
OPTIONS = -b -e
"))

(define gnunet-shepherd-service
  (match-lambda
    (($ <gnunet-configuration> package config-file)
     (list (shepherd-service
            (provision '(gnunet))
            ;; do we require networking? arm will try to reconnect until a connection
            ;; exists (again), but we might also set up vpn and not succeed at service
            ;; boot time as well as the general certificate issue we have especially on
            ;; Guix-on-GuixSD systems.
            (requirement '(loopback))
            (documentation "Run the GNUnet service.")
            (start
             (let ((gnunet
                    (file-append package "/lib/gnunet/libexec/gnunet-service-arm")))
               #~(make-forkexec-constructor
                  (list #$gnunet "-c" #$config-file)
                  #:log-file "/var/log/gnunet.log"
                  #:pid-file "/var/run/gnunet/arm-service.pid")))
            (stop
             #~(make-kill-destructor)))))))

(define %gnunet-accounts
  (list (user-group
         (name "gnunetdns")
         (system? #t))
        (user-group
         (name "gnunet")
         (system? #t))
        (user-account
         (name "gnunet")
         (group "gnunet")
         (system? #t)
         (comment "GNUnet system user")
         (home-directory "/var/lib/gnunet")
         (shell #~(string-append #$shadow "/sbin/nologin")))))

;; TODO: setuids.
;; TODO: certificate issues -- gnunet should honor CURL_CA_BUNDLE!
(define gnunet-activation
  (match-lambda
    (($ <gnunet-configuration> package config-file)
     (let ((gnunet
            (file-append package "/lib/gnunet/libexec/gnunet-service-arm")))
       #~(begin
           ;; Create the .config + .cache for gnunet user
           (mkdir-p "/var/lib/gnunet/.config/gnunet")
           (mkdir-p "/var/lib/gnunet/.cache/gnunet"))))))

(define gnunet-service-type
  (service-type
   (name 'gnunet)
   (extensions (list (service-extension account-service-type
                                        (const %gnunet-accounts))
                     (service-extension activation-service-type
                                        gnunet-activation)
                     (service-extension profile-service-type
                                        (compose list gnunet-configuration-package))
                     (service-extension shepherd-root-service-type
                                        gnunet-shepherd-service)))))

;;; gnunet.scm ends here
